pkg/controller/scyllaclustermigration/migrate_scyllacluster.go

```diff
@@ -0,0 +1,72 @@
+// Copyright (c) 2024 ScyllaDB.
+
+package scyllaclustermigration
+
+import (
+	"context"
+	"fmt"
+	"maps"
+
+	scyllav1 "github.com/scylladb/scylla-operator/pkg/api/scylla/v1"
+	"github.com/scylladb/scylla-operator/pkg/controller/scylladbdatacenter"
+	"github.com/scylladb/scylla-operator/pkg/controllerhelpers"
+	"github.com/scylladb/scylla-operator/pkg/naming"
+	"github.com/scylladb/scylla-operator/pkg/resourceapply"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func (scmc *Controller) syncScyllaDBDatacenter(ctx context.Context, sc *scyllav1.ScyllaCluster) ([]metav1.Condition, error) {
+	var progressingConditions []metav1.Condition
+
+	sdc, upgradeContext, err := migrateV1ScyllaClusterToV1Alpha1ScyllaDBDatacenter(sc)
+	if err != nil {
+		return progressingConditions, fmt.Errorf("can't migrate scyllav1.ScyllaCluster to v1alpha1.ScyllaDBDatacenter: %w", err)
+	}
+
+	existingSDc, err := scmc.scyllaDBDatacenterLister.ScyllaDBDatacenters(sc.Namespace).Get(sc.Name)
+	if err != nil && !errors.IsNotFound(err) {
+		return progressingConditions, fmt.Errorf("can't get v1alpha1.ScyllaDBDatacenter %q: %w", naming.ManualRef(sc.Namespace, sc.Name), err)
+	}
+
+	// Non-nil upgrade context means Operator was updated during upgrade of ScyllaCluster.
+	// Before we create the migrated ScyllaDBDatacenter, we have to make sure the ongoing upgrade state is migrated to new ConfigMap which holds it.
```

Comment:
What if we have already migrated the ConfigMap, then the operator was rolled back, the state has changed, then a new operator version was redeployed and the state on ScyllaCluster.status and the ConfigMap differ?

Reply:
Don't we need to migrate the state but block the new state machine, delete the old state and enable the new state machine?

Reply:
> What if we have already migrated the ConfigMap, then the operator was rolled back, the state has changed, then a new operator version was redeployed and the state on ScyllaCluster.status and the ConfigMap differ?

This scenario assumes there're no additional steps required during rollback. Since state machine authoritative source changes when Operator is rolled back, users would need to manually migrate it back to ScyllaCluster status. 

> Don't we need to migrate the state but block the new state machine, delete the old state and enable the new state machine?

We can't delete the old state, as currently it's visible to the user and it needs to stay like that.
Delaying of ScyllaDBDatacenter creation is essentially a pause of the new state machine. As part of rollback procedure, users have to orphan delete ScyllaDBDatacenter to allow dependants to be adopted by the ScyllaCluster, so blocking would prevent state skew in case of redeployment.